add a blog for addon rollout in Chinese (#387)
open-cluster-management-io/open-cluster-management-io.github.io: content/zh/blog/addon-rollout/_index.md

本文我们将介绍Open Cluster Management(OCM)如何将工具链升级视为配置文件的变更，使用户能够利用Kustomize或GitOps实现跨集群的无缝滚动/金丝雀升级。

在正式开始前，首先介绍几个OCM中的概念。

##  add-on插件

在OCM平台上，add-on插件可以实现不同被管理集群（Spoke）上应用的不同的配置，也可以实现从控制面（Hub）获取数据到Spoke集群上等功能。比如：你可以使用[managed-serviceaccount](https://github.com/open-cluster-management-io/managed-serviceaccount)
插件在Spoke集群上将指定的ServiceaCount信息返回给Hub集群，可以使用[cluster-proxy](https://github.com/open-cluster-management-io/cluster-proxy)插件建立一个从spoke到hub的反向代理通道。

现阶段OCM社区已经有的一些add-on：
* [Multicluster Mesh Addon](https://github.com/open-cluster-management-io/multicluster-mesh) 可用于管理（发现、部署和联合）OCM 中跨多个集群的服务网格。
* [Submarine Addon](https://github.com/stolostron/submariner-addon) 让[Submarine](https://github.com/submariner-io/submariner)
  和OCM方便集成，在hub cluster上部署Submariner Broker，在managed cluster上部署所需的Submariner组件, 为被管理集群提供跨集群的Pod和Service网络互相访问的能力。
* [Open-telemetry add-on](https://github.com/open-cluster-management-io/addon-contrib/tree/main/open-telemetry-addon) 自动在 hub cluster和managed cluster上

> 安装 otelCollector，并在 hub cluster上自动安装 jaeget-all-in-one 以处理和存储traces。

Reviewer: jaeger-all-in-one ?

Author: fixed

---
title: 以GitOps方式应对多集群工具链的升级挑战
---

2023年10月27日 [郝青](https://github.com/zhiweiyin318)

{{< toc >}}

# 多集群环境下工具链的升级挑战

OCM（open-cluster-management）是一个专注于Kubernetes应用跨多集群和多云的管理平台，提供了集群的注册，应用和负载的分发，调度等基础功能。Add-on插件是OCM提供的一种基于基础组件的扩展机制，可以让Kubernetes生态的应用很容易迁移到OCM平台上，拥有跨多集群多云的编排和调度的能力。如Istio，Prometheus，Submarine可以通过Add-on的方式扩展至多集群。在多集群环境中，如何优雅、平滑地升级整个工具链（比如Istio、Prometheus和其他工具）是我们在多集群管理中遇到的挑战，工具链的升级失败可能会导致数千个用户工作负载无法访问。因此，找到一种简单、安全的跨集群升级解决方案变得非常重要。

本文我们将介绍Open Cluster Management(OCM)如何将工具链升级视为配置文件的变更，使用户能够利用Kustomize或GitOps实现跨集群的无缝滚动/金丝雀升级。

在正式开始前，首先介绍几个OCM中的概念。

##  add-on插件

> 在OCM平台上，add-on插件可以实现不同被管理集群（Spoke）上应用的不同的配置，也可以实现从控制面（Hub）获取数据到Spoke集群上等功能。比如：你可以使用[managed-serviceaccount](https://github.com/open-cluster-management-io/managed-serviceaccount)

Reviewer 2: > add-on插件可以实现不同被管理集群（Spoke）上应用的不同的配置

读着有点不通顺

Author: fixed

* [Application lifecycle management](https://open-cluster-management.io/zh/getting-started/integration/app-lifecycle/) 
  实现多集群或多云环境中的应用程序生命周期管理。add-on插件提供了一套通过Subscriptions订阅channel，将github仓库，Helm release或者对象存储仓库的应用分发到指定Spoke集群上的机制。
* [Policy framework](https://open-cluster-management.io/getting-started/integration/policy-framework/) 和 
  [Policy controllers](https://open-cluster-management.io/getting-started/integration/policy-controllers/) 
  add-on插件可以让Hub集群管理员很轻松为Spoke集群部署安全相关的policy策略。
* [Managed service account](https://open-cluster-management.io/getting-started/integration/managed-serviceaccount/) 
  add-on插件可以让Hub集群管理员很容易管理Spoke集群上serviceaccount。
* [Cluster proxy](https://open-cluster-management.io/getting-started/integration/cluster-proxy/) add-on插件通过反向代理通道提供了Hub和Spoke集群之间L4网络连接。

**更多关于add-on插件的介绍可以参考[详解OCM add-on插件](https://open-cluster-management.io/zh/blog/addon-introduction/)。**

OCM提供了两种方式帮助开发者开发自己的add-on：
- Hard模式：使用[addon-framework](https://github.com/open-cluster-management-io/addon-framework)的内置机制，可根据[Add-on开发指南](https://open-cluster-management.io/developer-guides/addon/)来开发add-on插件的addon manager和addon agent。

> - Easy模式：OCM提供了一个新的插件开发模型，可使用[AddOnTemplate](https://open-cluster-management.io/developer-guides/addon/#build-an-addon-with-addon-template)来构建add-on。在此模型中开发者无需开发ddon manager，只需准备addon agent的image和AddOnTemplate，AddOnTemplate描述了如何部署addon agent以及如何注册addon。

Reviewer 2: ```suggestion
- Easy模式：OCM 提供了一个新的插件开发模型，可使用[AddOnTemplate](https://open-cluster-management.io/developer-guides/addon/#build-an-addon-with-addon-template)来构建 add-on。在此模型中开发者无需开发 addon manager，只需准备 addon agent 的 image 和AddOnTemplate，AddOnTemplate 描述了如何部署 addon agent 以及如何注册 addon。
```

Reviewer 2: addon manager 少了开头的 a

Author: fixed

spec:
  addOnMeta:
    description: hello-template is a addon built with addon template
    displayName: hello-template
  supportedConfigs: # declare it is a template type addon
  - group: addon.open-cluster-management.io
    resource: addontemplates
    defaultConfig:
      name: hello-template
```

```yaml
apiVersion: addon.open-cluster-management.io/v1alpha1
kind: AddOnTemplate
metadata:
  name: hello-template
spec:
  addonName: hello-template
  agentSpec: # required
      workload:
        manifests:
          - kind: Deployment
            metadata:
              name: hello-template-agent
              namespace: open-cluster-management-agent-addon
...
          - kind: ServiceAccount
            metadata:
              name: hello-template-agent-sa
              namespace: open-cluster-management-agent-addon
          - kind: ClusterRoleBinding
             metadata:
               name: hello-template-agent
...
  registration: # optional
    ...
```

## Placement Decision Strategy

Placement API用于在一个或多个托管集群组(ManagedClusterSet)中选择一组托管群集(ManagedCluster)，以便将工作负载部署到这些群集上。

**更多关于Placement API的介绍可以参考[Placement文档](https://open-cluster-management.io/concepts/placement/)。**

> Placement调度过程的“输入”和“输出”被解耦为两个独立的 Kubernetes API Placement 和 PlacementDecision.

Reviewer 2: ```suggestion
Placement 调度过程的“输入”和“输出”被解耦为两个独立的 Kubernetes API: Placement 和 PlacementDecision。
```

    group: addon.open-cluster-management.io
    resource: addontemplates
  installStrategy:
    placements:
    - name: aws-placement
      namespace: default
      configs:
      - group: addon.open-cluster-management.io
        resource: addondeploymentconfigs
        name: managed-serviceaccount-addon-deploy-config
      rolloutStrategy:
        type: Progressive
        progressive:
          mandatoryDecisionGroups:
          - groupName: "canary"
          maxConcurrency: 1
    type: Placements
```

- 版本控制

add-on配置文件名称或spec内容的变化会被认为是一个配置更改，会触发add-on的一次升级。用户可以利用Kustomize或GitOps来控制配置文件升级。

```yaml
apiVersion: addon.open-cluster-management.io/v1alpha1
kind: AddOnTemplate
metadata:
  name: managed-serviceaccount-0.4.0
spec:
  agentSpec: # required
      workload:
        manifests:
          - kind: Deployment
            metadata:
              name: managed-serviceaccount-addon-agent
              namespace: open-cluster-management-agent-addon
..
          - kind: ServiceAccount
            metadata:
              name: managed-serviceaccount
              namespace: open-cluster-management-agent-addon
…
  registration: # optional
```

- 自动化

> OCM在open-cluster-management-hub命名空间下的组件addon-manager-controller是一个更通用的addon manager，它会watch使用AddOnTemplate模式开发的add-on并负责维护此类add-on的生命周期，包括安装与升级。当配置文件的名称或者spec内容变化时，此组件会按照rolloutStrategy所定义的升级策略来升级add-on。

Reviewer 2: 也会watch `addon.open-cluster-management.io/lifecycle: "addon-manager"`  ?

Author: 加上了

```bash

> ✗ kubectl  get deploy -n open-cluster-management-hub

Reviewer 2: ```suggestion
✗ kubectl get deploy -n open-cluster-management-hub
```

Author: fixed

---
title: 以GitOps方式应对多集群工具链的升级挑战

> title: 使用OCM滚动升级多集群下的工具链

Author: @qiujian16 any suggestion on the title?

Reviewer: I prefer the 1st one

安装 otelCollector，并在 hub cluster上自动安装 jaeget-all-in-one 以处理和存储traces。
* [Application lifecycle management](https://open-cluster-management.io/zh/getting-started/integration/app-lifecycle/) 
  实现多集群或多云环境中的应用程序生命周期管理。add-on插件提供了一套通过Subscriptions订阅channel，将github仓库，Helm release或者对象存储仓库的应用分发到指定Spoke集群上的机制。
* [Policy framework](https://open-cluster-management.io/getting-started/integration/policy-framework/) 和 
  [Policy controllers](https://open-cluster-management.io/getting-started/integration/policy-controllers/) 
  add-on插件可以让Hub集群管理员很轻松为Spoke集群部署安全相关的policy策略。
* [Managed service account](https://open-cluster-management.io/getting-started/integration/managed-serviceaccount/) 
  add-on插件可以让Hub集群管理员很容易管理Spoke集群上serviceaccount。
* [Cluster proxy](https://open-cluster-management.io/getting-started/integration/cluster-proxy/) add-on插件通过反向代理通道提供了Hub和Spoke集群之间L4网络连接。

OCM提供了两种方式帮助开发者开发自己的add-on：
- hard模式：使用[addon-framework](https://github.com/open-cluster-management-io/addon-framework)的内置机制，可根据[Add-on开发指南](https://open-cluster-management.io/developer-guides/addon/)来开发add-on插件的addon manager和addon agent。

> - easy模式：OCM提供了一个新的插件开发模型，可使用[AddOnTemplate](https://open-cluster-management.io/developer-guides/addon/#build-an-addon-with-addon-template)来构建add-on。在此模型中开发者无需开发ddon manager，只需准备addon agent的image和AddOnTemplate，AddOnTemplate描述了如何部署addon agent以及如何注册addon。

Reviewer 3: nil: Hard,Easy

Author: fixed

```yaml
apiVersion: cluster.open-cluster-management.io/v1beta1
kind: PlacementDecision
metadata:
  labels:
    cluster.open-cluster-management.io/decision-group-index: "0"
    cluster.open-cluster-management.io/decision-group-name: canary
    cluster.open-cluster-management.io/placement: aws-placement
  name: aws-placement-decision-1
  namespace: default
status:
  decisions:
  - clusterName: cluster1
    reason: ""
...
  - clusterName: cluster10
    reason: ""
```

## 以GitOps方式简化升级
以上简单介绍了add-on template和placement decision strategy的概念。更多详细的介绍可访问我们的[官网](https://open-cluster-management.io/concepts/)了解。

在OCM中，我们将add-on的升级视为其配置文件的升级，这里的配置可以是AddOnTemplate，也可以是其他自定义的配置文件AddOnDeploymentConfig。一次add-on的升级等同于一次配置文件的更新，这使得用户能够利用Kustomize或GitOps来进行无缝的跨集群滚动/金丝雀升级。RolloutStrategy定义了升级策略，支持全部升级(All)，按集群渐进升级(Progressive Per Cluster)和按集群组渐进升级(Progressive Per Group)，并可定义一组MandatoryDecisionGroups来优先尝试新配置。

依照GitOps的四个原则，我们来看看OCM如何支持以GitOps的方式应对多集群环境下的升级挑战。

- 声明式

> 在`ClusterManagementAddOn`中可以声明addo-on所使用的配置文件。配置文件可在全局的`supportedConfigs`中声明，该配置文件会应用到所有的`ManagedClusterAddOn`实例上。也可在`installStrategy`下不同的placements中声明，每个Placement所选择集群的`ManagedClusterAddOn`将拥有相同的配置文件，placements中声明的配置会覆盖全局配置。

Reviewer 3: typo： add-on

Author: fixed
